### ISSUE TITLE
Different templates with different blocks but same block names

### ISSUE THREAD

**Issue by michaelzangerle** [sulu/sulu#5983]
| Q | A
| --- | ---
| Sulu Version | current master

First of all I am not sure this is a bug or just a missing feature or something else. As I ran into it because of an error I chose the bug template to create this issue. I talked to @wachterjohannes and we decided to create an issue to discuss it here because we uncovered some more scenarios as well.

**Situation** 
- we have two templates T1 and T2
- we have blocks `headline`, `products`, `products-automatic`, `teaser`

# Actual Behavior
While switching from T1 with a configured `headline` to T2 I ran into the following error `The block type "headline" has not been registered. Known block types are: [products, products-automatic, teaser]`

# Expected Behavior
I was expecting / we were discussing the following scenarios:
- if `headline` is only registered in T1 but not in T2 it shouldn't have any effect on the new template and get removed
- if `headline` is registered in both templates and has the same definition it should be possible to kept the information
- if `headline` is registered in both templates and they have at least a partially compatible definition as much as possible information should be kept
- if `headline` is registered in both templates but the have incompatible definitions no information should be kept

Those were the things we discussed very briefly and it seems like some of them are already implemented but not all of them. Is this something you want to do? If not what other solution would you suggest to prevent this error from happening? Prefixing block names in templates?

**Comment by niklasnatter** [comment]
Hey, thanks for reporting!
We talked about this today and decided that the frontend should fallback to the default type if a block-type is not available in the current template. The data of the block should not be changed and therefore matching properties should keep the data. This behaviour should be consistent with the behaviour for properties outside of blocks 🙂

**Comment by michaelzangerle** [comment]
@nnatter That sounds good. Would this `decided that the frontend should fallback to the default type` also mean that the invalid data wont be sent to the backend in the first place and not cause this exception?

```
InvalidArgumentException: The block type "teaser-builder" has not been registered. Known block types are: [products, products-automatic, teaser]
#19 /vendor/sulu/sulu/src/Sulu/Component/Content/Compat/Property.php(514): Sulu\Component\Content\Compat\Property::getType
#18 /vendor/sulu/sulu/src/Sulu/Component/Content/Compat/Property.php(550): Sulu\Component\Content\Compat\Property::initProperties
#17 /vendor/sulu/sulu/src/Sulu/Component/Content/Compat/Block/BlockProperty.php(97): Sulu\Component\Content\Compat\Block\BlockProperty::doSetValue
#16 /vendor/sulu/sulu/src/Sulu/Component/Content/Compat/Block/BlockProperty.php(68): Sulu\Component\Content\Compat\Block\BlockProperty::setPropertyValue
#15 /vendor/sulu/sulu/src/Sulu/Component/Content/Compat/Structure/StructureBridge.php(674): Sulu\Component\Content\Compat\Structure\StructureBridge::createLegacyPropertyFromItem
#14 /vendor/sulu/sulu/src/Sulu/Component/Content/Compat/Structure/StructureBridge.php(201): Sulu\Component\Content\Compat\Structure\StructureBridge::getProperties
#13 /vendor/sulu/headless-bundle/Content/StructureResolver.php(78): Sulu\Bundle\HeadlessBundle\Content\StructureResolver::resolve
#12 /vendor/sulu/headless-bundle/Controller/HeadlessWebsiteController.php(62): Sulu\Bundle\HeadlessBundle\Controller\HeadlessWebsiteController::resolveStructure
#11 /vendor/sulu/headless-bundle/Controller/HeadlessWebsiteController.php(37): Sulu\Bundle\HeadlessBundle\Controller\HeadlessWebsiteController::indexAction
#10 /vendor/symfony/http-kernel/HttpKernel.php(157): Symfony\Component\HttpKernel\HttpKernel::handleRaw
#9 /vendor/symfony/http-kernel/HttpKernel.php(79): Symfony\Component\HttpKernel\HttpKernel::handle
#8 /vendor/symfony/http-kernel/Kernel.php(195): Symfony\Component\HttpKernel\Kernel::handle
#7 /vendor/sulu/sulu/src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php(216): Sulu\Bundle\PreviewBundle\Preview\Renderer\PreviewRenderer::handle
#6 /vendor/sulu/sulu/src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php(186): Sulu\Bundle\PreviewBundle\Preview\Renderer\PreviewRenderer::render
#5 /vendor/sulu/sulu/src/Sulu/Bundle/PreviewBundle/Preview/Preview.php(128): Sulu\Bundle\PreviewBundle\Preview\Preview::updateContext
#4 /vendor/sulu/sulu/src/Sulu/Bundle/PreviewBundle/Controller/PreviewController.php(119): Sulu\Bundle\PreviewBundle\Controller\PreviewController::updateContextAction
#3 /vendor/symfony/http-kernel/HttpKernel.php(157): Symfony\Component\HttpKernel\HttpKernel::handleRaw
#2 /vendor/symfony/http-kernel/HttpKernel.php(79): Symfony\Component\HttpKernel\HttpKernel::handle
#1 /vendor/symfony/http-kernel/Kernel.php(195): Symfony\Component\HttpKernel\Kernel::handle
#0 /public/index.php(66): null
Sulu\Bundle\PreviewBundle\Preview\Exception\TemplateNotFoundException: The block type "teaser-builder" has not been registered. Known block types are: [products, products-automatic, teaser]
#6 /vendor/sulu/sulu/src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php(196): Sulu\Bundle\PreviewBundle\Preview\Renderer\PreviewRenderer::render
#5 /vendor/sulu/sulu/src/Sulu/Bundle/PreviewBundle/Preview/Preview.php(128): Sulu\Bundle\PreviewBundle\Preview\Preview::updateContext
#4 /vendor/sulu/sulu/src/Sulu/Bundle/PreviewBundle/Controller/PreviewController.php(119): Sulu\Bundle\PreviewBundle\Controller\PreviewController::updateContextAction
#3 /vendor/symfony/http-kernel/HttpKernel.php(157): Symfony\Component\HttpKernel\HttpKernel::handleRaw
#2 /vendor/symfony/http-kernel/HttpKernel.php(79): Symfony\Component\HttpKernel\HttpKernel::handle
#1 /vendor/symfony/http-kernel/Kernel.php(195): Symfony\Component\HttpKernel\Kernel::handle
#0 /public/index.php(66): null
```

**Comment by niklasnatter** [comment]
Yes, the frontend should send the type that is displayed to the user to the server 🙂

**Comment by alexander-schranz** [comment]
I did test this how this was handled in the `1.6` and it seems like in `1.6` the page data was loaded from the backend with the new template. There you see if you switch the template the page is loaded again from the backend with:

>  /admin/api/nodes/8666fd70-0c0b-49ee-8848-027acca087bd?webspace=example&language=en&breadcrumb=true&template=homepage

So it is loaded with the supported block types. The problem in 1.6 was when you did switch the template in the `Add Form` all your inputed data was getting lost. So not sure if this is the expected behaviour.

**Comment by niklasnatter** [comment]
Thanks for reporting this! It should be fixed by #6077 in the upcoming Sulu 2.3.1 release 🙂
